docs: update architecture doc to introduce the `containerd-shimkit` crate (#933)
containerd/runwasi: docs/src/developer/architecture.md

## High-Level Architecture

Runwasi s designed as a library that can be integrated with WebAssembly runtimes to enable them to be used with containerd. The following diagram illustrates the high-level architecture:

![A diagram of runwasi architecture](../assets/runwasi-architecture.png)

## Project Structure

The Runwasi project is organized into several components:

> - **containerd-shimkit** - A lower level, opinionated library providing a API for building containerd shims. This serves as the lower layer for all shim implementations.
> - **containerd-shim-wasm** - A higher level library that is used by WebAssembly runtimes to create shims, built on top of `containerd-shimkit`. Most of the WebAssembly-specific shared code lives here.
> - **containerd-shim-wasm-test-modules** - Library with WebAssembly test modules used in the testing framework.
> - **containerd-shim-\<runtime>** - Shim reference implementation for selected runtimes (wasmtime, wasmedge, wasmer, wamr, etc.). These produce binaries that are the shims which containerd can communicate with.
> - **oci-tar-builder** - Library and executable that helps build OCI tar files that follow the [`wasm-oci` spec](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).

Reviewer: **suggestion:** De-emphasize `containerd-shimkit` crate by moving it further down, and a bit of rewording
```suggestion
- **containerd-shim-wasm** - A higher level library that is used by WebAssembly runtimes to create shims. Most of the WebAssembly-specific shared code lives here.
- **containerd-shim-wasm-test-modules** - Library with WebAssembly test modules used in the testing framework.
- **containerd-shim-\<runtime>** - Shim reference implementation for selected runtimes (wasmtime, wasmedge, wasmer, wamr, etc.). These produce binaries that are the shims which containerd can communicate with.
- **containerd-shimkit** - A lower level, opinionated library providing a API for building containerd shims. It serves as the building block of `containerd-shim-wasm`.
- **oci-tar-builder** - Library and executable that helps build OCI tar files that follow the [`wasm-oci` spec](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).
```

The Containerd "shim" is a daemon process that serves the [Task Service API](https://github.com/containerd/containerd/blob/v2.0.0/core/runtime/v2/README.md#tasks). It listens on a socket to receive ttrpc requests from Containerd, allowing for lifecycle management for containers (create, start, stop, etc.). 

### Runwasi Libraries

> The core of Runwasi consists of two main Rust libraries:

Reviewer: **suggestion:** I would de-emphasize containerd-shimkit as we are not planning to stabilize it.
```suggestion
The core of Runwasi is the `containerd-wasm-shim` library crate.
```

> #### Low-level: `containerd-shimkit`
> 
> The `containerd-shimkit` crate provides a low-level API such as the `Instance` trait for implementing a shim. It supports both MacOS and Windows and provides observability and tracing for monitoring shim operations. However, it also has limitations:
> 
> - No precompilation out-of-the-box
> - Does not support for native Linux containers out-of-the-box
> - Requires manual handling of cgroup setup
> 
> #### High-level: `containerd-shim-wasm`
> 
> Built on top of containerd-shimkit, the `containerd-shim-wasm` crate provides a high-level API for building WebAssembly shims, mainly the `Engine` trait. The `Engine` trait has the following features:
> 
> 1. **Shim Implementation**: Because it is built on top of `containerd-shimkit`, it implements the containerd shim v2 API to facilitate communication between containerd and the WebAssembly runtime.
> 2. **Wasm OCI Integration**: Transparent handling of the [wasm-oci spec](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).
> 3. **Wasm-specific Features**: Support for Wasm module or component validation and precompilation.

Reviewer: **suggestion:** I would de-emphasize containerd-shimkit as we are not planning to stabilize it.
```suggestion
The `containerd-shim-wasm` crate provides a high-level API for building WebAssembly shims, mainly the `Engine` trait. The `Engine` trait has the following features:

1. **Shim Implementation**: It implements the containerd shim v2 API to facilitate communication between containerd and the WebAssembly runtime. This is done through the `containerd-shimkit`'s `Instance` trait (more on this trait below).
2. **Wasm OCI Integration**: Transparent handling of the [wasm-oci spec](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).
3. **Wasm-specific Features**: Support for Wasm module or component validation and precompilation.

The `Engine` trait is build on top of the `containerd-shimkit`'s `Instance` trait. The `Instance` trait provides a low-level API for implementing containerd shims. It supports Linux, MacOS, and Windows and provides observability and tracing for monitoring shim operations. However, it also has limitations:

- It is *not* stable, and considered an implementation detail for `containerd-shim-wasm`
- No precompilation out-of-the-box
- Does not support for native containers out-of-the-box
- Requires manual handling of cgroup setup
```

### Runwasi Libraries

The core of Runwasi is the `containerd-wasm-shim` library crate.

The `containerd-shim-wasm` crate provides a high-level API for building WebAssembly shims, mainly the `Engine` trait. The `Engine` trait has the following features:

1. **Shim Implementation**: It implements the containerd shim v2 API to facilitate communication between containerd and the WebAssembly runtime. This is done through the `containerd-shimkit`'s `Instance` trait (more on this trait below).
2. **Wasm OCI Integration**: Transparent handling of the [wasm-oci spec](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).
3. **Wasm-specific Features**: Support for Wasm module or component validation and precompilation.

> The `Engine` trait is build on top of the `containerd-shimkit`'s `Instance` trait. The `Instance` trait provides a low-level API for implementing containerd shims. It supports Linux, MacOS, and Windows and provides observability and tracing for monitoring shim operations. However, it also has limitations:

Reviewer: ```suggestion
The `Engine` trait is built on top of the `containerd-shimkit`'s `Instance` trait. The `Instance` trait provides a low-level API for implementing containerd shims. It supports Linux, MacOS, and Windows and provides observability and tracing for monitoring shim operations. However, it also has limitations:
```
